### PR TITLE
[rnn/gru] : no batch dim

### LINKED ISSUE #60585: Rollup: No-batch-dim support for torch.nn modules
## Background
A previous issue (https://github.com/pytorch/pytorch/issues/47149) requested support of arbitrary batch dimensions across module inputs. This rollup issue details the `torch.nn` module updates required to address a subset of this functionality: specifically, the case of no batch dimensions. This particular case is useful for composability with a future vmap.

## Module support
- [x] `nn.AdaptiveAvgPool1d` - Should be updated to support 2D inputs
- [x] `nn.AdaptiveAvgPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveAvgPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveLogSoftmaxWithLoss` - Should be updated to support 1D input and scalar target / output1 (@george-qi)
- [x] `nn.AdaptiveMaxPool1d` - Should be updated to support 2D inputs
- [x] `nn.AdaptiveMaxPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveMaxPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.AlphaDropout`
- [x] `nn.AvgPool1d` - Should be updated to support 2D inputs
- [x] `nn.AvgPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AvgPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.BCELoss`
- [x] `nn.BCEWithLogitsLoss`
- [x] `nn.Bilinear` - Already supports no batch dims, but docs should make this clear (@george-qi)
- [x] `nn.CELU`
- [x] `nn.CTCLoss` - Should be updated to support a single (source, target) sequence (@george-qi)
- [x] `nn.ConstantPad1d`
- [x] `nn.ConstantPad2d`
- [x] `nn.ConstantPad3d`
- [x] `nn.Conv1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.Conv2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.Conv3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.ConvTranspose1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.ConvTranspose2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.ConvTranspose3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.CosineEmbeddingLoss` - Should be updated to support 1D input with scalar target (@kshitij12345 )
- [x] `nn.CosineSimilarity` (pass `dim=0` for no-batch-dim support)
- [x] `nn.CrossEntropyLoss` - Should be updated to support 1D input with scalar target (@kshitij12345 )
- [x] `nn.Dropout`
- [x] `nn.Dropout2d`
- [x] `nn.Dropout3d`
- [x] `nn.ELU`
- [x] `nn.Embedding`
- [x] `nn.FeatureAlphaDropout` - Already supports no batch dims, but docs are missing
- [x] `nn.Flatten` (pass `start_dim=0` for no-batch-dim support)
- [x] `nn.Fold` - Should be updated to support 3D inputs (@kshitij12345)
- [x] `nn.FractionalMaxPool2d` - Should be updated to support 3D inputs
- [x] `nn.FractionalMaxPool3d` - Should be updated to support 4D inputs (@george-qi)
- [x] `nn.GELU`
- [x] `nn.GLU`
- [x] `nn.GRU` - Should be updated to support a single input sequence / hidden state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.GRUCell` - Should be updated to support 1D input / hidden state (@kshitij12345)
- [x] `nn.GaussianNLLLoss` - Already supports no batch dims, but docs should make this clear (@george-qi)
- [x] `nn.Hardshrink`
- [x] `nn.Hardsigmoid`
- [x] `nn.Hardswish`
- [x] `nn.Hardtanh`
- [x] `nn.HingeEmbeddingLoss`
- [x] `nn.HuberLoss`
- [x] `nn.Identity`
- [x] `nn.InstanceNorm1d` - Should be updated to support 1D or 2D inputs (@kshitij12345)
- [x] `nn.InstanceNorm2d` - Should be updated to support 2D or 3D inputs (@kshitij12345)
- [x] `nn.InstanceNorm3d` - Should be updated to support 3D or 4D inputs (@kshitij12345)
- [x] `nn.KLDivLoss`
- [x] `nn.L1Loss`
- [x] `nn.LPPool1d` - Should be updated to support 2D inputs
- [x] `nn.LPPool2d`
- [x] `nn.LSTM` - Should be updated to support single input, hidden state, cell state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.LSTMCell` - Should be updated to support 1D input, hidden state, cell state (@kshitij12345)
- [x] `nn.LayerNorm` (pass correctly-set `normalized_shape` for no-batch-dim support)
- [x] `nn.LazyConv1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.LazyConv2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.LazyConv3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.LazyConvTranspose1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.LazyConvTranspose2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.LazyConvTranspose3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.LazyLinear` - Already supports no batch dims, but docs should make this clear
- [x] `nn.LeakyReLU`
- [x] `nn.Linear` - Already supports no batch dims, but docs should make this clear
- [x] `nn.LogSigmoid`
- [x] `nn.LogSoftmax`
- [x] `nn.MSELoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.MarginRankingLoss` - Should be updated to support scalars (@kshitij12345)
- [x] `nn.MaxPool1d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxPool2d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxPool3d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxUnpool1d` - Should be updated to support 2D input / indices
- [x] `nn.MaxUnpool2d` - Should be updated to support 3D input / indices
- [x] `nn.MaxUnpool3d` - Should be updated to support 4D input / indices
- [x] `nn.Mish`
- [x] `nn.MultiLabelMarginLoss`
- [x] `nn.MultiLabelSoftMarginLoss` - Should be updated to support 1D input / target
- [x] `nn.MultiMarginLoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.MultiheadAttention` - Should be updated to support unbatched query, key, values, and masks; `batch_first` arg is meaningless in this case
- [x] `nn.NLLLoss` - Should be updated to support 1D input and scalar target
- [x] `nn.PairwiseDistance` - Should be updated to support 1D inputs with scalar output
- [x] `nn.PixelShuffle`
- [x] `nn.PixelUnshuffle`
- [x] `nn.PoissonNLLLoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.RNN` - Should be updated to support a single input sequence / hidden state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.RNNCell` - Should be updated to support 1D input and hidden state (@kshitij12345)
- [x] `nn.RReLU`
- [x] `nn.ReLU`
- [x] `nn.ReLU6`
- [x] `nn.ReflectionPad1d` - Should be updated to support 2D inputs
- [x] `nn.ReflectionPad2d`
- [x] `nn.ReflectionPad3d`
- [x] `nn.ReplicationPad1d` - Should be updated to support 2D inputs
- [x] `nn.ReplicationPad2d`
- [x] `nn.ReplicationPad3d`
- [x] `nn.SELU`
- [x] `nn.SiLU`
- [x] `nn.Sigmoid`
- [x] `nn.SmoothL1Loss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.SoftMarginLoss`
- [x] `nn.Softmax`
- [x] `nn.Softmax2d` - Should be updated to support 3D input
- [x] `nn.Softmin`
- [x] `nn.Softplus`
- [x] `nn.Softshrink`
- [x] `nn.Softsign`
- [x] `nn.Tanh`
- [x] `nn.Tanhshrink`
- [x] `nn.Threshold`
- [x] `nn.Transformer` - Should be updated to support unbatched source / target sequences and masks (@kshitij12345)
- [x] `nn.TransformerDecoderLayer` - Should be updated to support unbatched inputs (@kshitij12345)
- [x] `nn.TransformerEncoderLayer` - Should be updated to support unbatched inputs (@kshitij12345)
- [x] `nn.TripletMarginLoss` - Should be updated to support 1D anchor, positive, and negative (@kshitij12345)
- [x] `nn.TripletMarginWithDistanceLoss` (support is based on the choice of distance function)
- [x] `nn.Unflatten` - Already supports no batch dims, but docs should make this clear
- [x] `nn.ZeroPad2d`

## Semantically incompatible modules
* `nn.BatchNorm1d` - Only defined over batch
* `nn.BatchNorm2d` - Only defined over batch
* `nn.BatchNorm3d` - Only defined over batch
* `nn.LazyBatchNorm1d` - Only defined over batch
* `nn.LazyBatchNorm2d` - Only defined over batch
* `nn.LazyBatchNorm3d` - Only defined over batch
* `nn.SyncBatchNorm` - Only defined over batch

## Modules that require BC-breaking changes
* `nn.ChannelShuffle` - While the docs say that the module accepts shape (*, C, H, W), the implementation assumes shape (N, C, *) with * being 1 or more dimensions; switching to (C, *) is BC-breaking because it would reinterpret dims
* `nn.EmbeddingBag` - Already supports 1D inputs with different semantics
* `nn.GroupNorm` - Supports arbitrary spatial dims; switching to (C, *) instead of (N, C, *) is BC-breaking since it would reinterpret dims
* `nn.LocalResponseNorm` - Supports (N, C, *) shape now; switching to (C, *) is BC-breaking since it would reinterpret dims
* `nn.PReLU` - Always assumes the 2nd dim is the channels dim
* `nn.Unfold` - Supports 4D inputs; switching to (C, *) instead of (N, C, *) is BC-breaking since it would reinterpret dims
* `nn.Upsample` - Supporting unbatched inputs would be BC-breaking because it would reinterpret dims

## Irrelevant modules
* `nn.Module`
* `nn.ModuleDict`
* `nn.ModuleList`
* `nn.ParameterDict`
* `nn.ParameterList`
* `nn.Sequential`
* `nn.TransformerDecoder` - see `nn.TransformerDecoderLayer` instead
* `nn.TransformerEncoder` - see `nn.TransformerEncoderLayer` instead

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #60585

TODO:
* [x] Doc updates

### CODE REVIEW

**Comment by kshitij12345** [comment]
I am not sure at this point if we should also look at altering the packed_sequence path (it does support unbatched input naturally as it takes care of batching but the output shape is different for hidden)
```python
import torch
from torch.nn.utils.rnn import pack_sequence
t = torch.randn(2, 2)

rnn = torch.nn.RNN(2, 2, num_layers=2)

o_ub = rnn(t)

o = rnn(pack_sequence([t]))

torch.testing.assert_allclose(o[0].data, o_ub[0]) # Output
torch.testing.assert_allclose(o[1].squeeze(1), o_ub[1]) # Hidden
```

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -429,6 +436,16 @@ def forward(self, input, hx=None):  # noqa: F811
             max_batch_size = int(batch_sizes[0])
         else:
             batch_sizes = None
+            is_batched = input.dim() == 3
+            batch_dim = 0 if self.batch_first else 1
+            if not is_batched:
+                input = input.unsqueeze(batch_dim)
+                if hx is not None:
+                    assert hx.dim() == 2
```
Comment (jbschlosser) [inline torch/nn/modules/rnn.py:444]: I think raising a `RuntimeError` is preferred to asserting, and I like the error messages you created for the `GRU` case - can we get similar ones here as well?

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -848,6 +877,18 @@ def forward(self, input, hx=None):  # noqa: F811
             max_batch_size = int(max_batch_size)
         else:
             batch_sizes = None
+            is_batched = input.dim() == 3
+            batch_dim = 0 if self.batch_first else 1
+            if not is_batched:
+                input = input.unsqueeze(batch_dim)
+                if hx is not None:
+                    assert hx.dim() == 2, \
```
Comment (jbschlosser) [inline torch/nn/modules/rnn.py:885]: Similarly here and below, a `RuntimeError` should be thrown instead of asserting

File: torch/testing/_internal/common_modules.py [hunk 3]
```diff
@@ -350,6 +351,27 @@ def no_batch_dim_reference_mha(m, p, *args, **kwargs):
         return (output[0].squeeze(batch_dim), output[1].squeeze(0))
 
 
+def no_batch_dim_reference_rnn_gru(m, p, *args, **kwargs):
+    """Reference function for RNN supporting no batch dimensions.
+    The module is passed the input and target in batched form with a single item.
+    The output is squeezed to compare with the no-batch input.
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:357]: Is this description correct? IIUC the input and target are passed to this function in unbatched form and unsqueezed to create a single item batch, then they are run through the module, then the output is squeezed to compare with the no-batch *output*

File: torch/testing/_internal/common_modules.py [hunk 4]
```diff
@@ -956,4 +1025,57 @@ def module_inputs_torch_nn_LSTMCell(module_info, device, dtype, requires_grad, *
...
+                   # CUDNN RNN doesn't accept non-contiguous hx
+                   DecorateInfo(
+                       unittest.expectedFailure, "TestModule", "test_non_contiguous_tensors",
+                       active_if=(TEST_CUDNN and not TEST_WITH_ROCM)
+                   ),
+                   # MIOPEN RNN doesn't accept non-contiguous hx (this is dispatched to miopen only for float).
+                   DecorateInfo(
+                       unittest.expectedFailure, "TestModule", "test_non_contiguous_tensors",
+                       active_if=(TEST_CUDNN and TEST_WITH_ROCM), dtypes=(torch.float,)
+                   ),
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:1052]: phew what a pain this must have been! sorry about all these failures - we'll have to address these at some point. Thanks for the descriptive comments for each!

**Review by jbschlosser — COMMENTED** [review]
Looking pretty good, just a few minor comments :) Thanks for the update!

File: torch/testing/_internal/common_modules.py (hunk 3 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_modules.py]: Have updated it (and the others which had same comment). Maybe I haven't phrased it nicely. PTAL :)

**Review by jbschlosser — APPROVED** [review]
LGTM! :D

**Comment by suo** [comment]
Looks like this broke windows test on master with stuff like:
```
2022-01-07T04:20:11.8667776Z   test_non_contiguous_tensors_nn_RNN_cpu_float32 (__main__.TestModuleCPU) ... unexpected success (0.732s)
2022-01-07T04:20:12.5086543Z   test_non_contiguous_tensors_nn_RNN_cpu_float64 (__main__.TestModuleCPU) ... unexpected success (0.638s)
```
https://www.torch-ci.com/commit/6eba936082a641be8ece156f70c0f5c435f7a7aa

I am reverting

**Comment by suo** [comment]
too successful! what a world
